src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,216 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public void linkLast(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
+    }
+
+    public void linkFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            head.prev = newNode;
+            head = newNode;
+        }
+        size++;
+    }
+
+    public void linkByIndex(T value, int index) {
+        indexCheck(index);
+
+        if (index == 0) {
+            linkFirst(value);
+        } else if (index == size) {
+            linkLast(value);
+        } else {
+            Node<T> nodeByIndex = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
+    }
+
+    public Node<T> findNodeByIndex(int index) {
+        indexCheck(index);
+        Node<T> currentNode;
+
+        if (index < size >> 1) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    public void indexCheck(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(
+                    "indexCheckException: index: " + index
+                            + " size: " + size);
+        }
+    }
+
+    public void indexEqualsSizeCheck(int index) {
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Can`t get element when index == size");
+        }
+    }
```

Comment:
The `indexEqualsSizeCheck` method is redundant and should be removed. Operations like `add` and `get` should handle the case where the index equals the size appropriately without this check.